source/fab/build_config.py

```diff
@@ -0,0 +1,199 @@
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+##############################################################################
+import getpass
+import logging
+import os
+from datetime import datetime
+from fnmatch import fnmatch
+from logging.handlers import RotatingFileHandler
+from multiprocessing import cpu_count
+from pathlib import Path
+from string import Template
+from typing import List
+
+from fab.constants import BUILD_OUTPUT, SOURCE_ROOT
+from fab.dep_tree import by_type
+from fab.metrics import send_metric, init_metrics, stop_metrics, metrics_summary
+from fab.steps import Step
+from fab.util import TimerLogger
+
+logger = logging.getLogger(__name__)
+
+
+class BuildConfig(object):
+    """
+    Contains and runs a list of build steps.
+
+    :param str project_label:
+        Name of the build project. The project workspace folder is created from this name, with spaces replaced
+        by underscores.
+    :param Path source_root:
+        Optional argument to allow the config to find the source code outside it's project workspace.
+        This is useful, for example, when the :py:mod:`fab.steps.grab <grab>` is in a separate script to be run
+        less frequently. In this scenario, the source code will be found in a  different project workspace folder.
+    :param List[Step] steps:
+        The list of build steps to run.
+    :param bool multiprocessing:
+        An option to disable multiprocessing to aid debugging.
+    :param int n_procs:
+        The number of cores to use for multiprocessing operations. Defaults to the number of available cores - 1.
+    :param bool reuse_artefacts:
+        A flag to avoid reprocessing artefacts a second time.
+        Currently unsophisticated, the existence of an output file means it doesn't need to be reprocessed.
+        The logic behind flag will be improved in the future.
+
+    """
+
+    def __init__(self, project_label, source_root=None, steps: List[Step] = None,
+                 multiprocessing=True, n_procs=None, reuse_artefacts=False):
+
+        self.project_label: str = project_label
+
+        # workspace folder
+        if os.getenv("FAB_WORKSPACE"):
+            fab_workspace_root = Path(os.getenv("FAB_WORKSPACE"))  # type: ignore
+        else:
+            fab_workspace_root = Path(os.path.expanduser("~/fab-workspace"))
+            logger.info(f"FAB_WORKSPACE not set. Using {fab_workspace_root}.")
+        self.project_workspace = fab_workspace_root / (project_label.replace(' ', '-'))
+        self.metrics_folder = self.project_workspace / 'metrics' / self.project_label.replace(' ', '_', -1)
+
+        # source config
+        self.source_root = source_root or self.project_workspace / SOURCE_ROOT
+
+        # build steps
+        self.steps: List[Step] = steps or []
```

Comment:
I think that's moot because we must [avoid default mutable arguments](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments). Sequence is more correct than List through.

Reply:
Oh, I went to change the argument type to `typing.Sequence` but it appears to be [deprecated](https://docs.python.org/3/library/typing.html#typing.Sequence) from 3.9. I think we _could_ get round that if we're using 3.7 minimum (which we're not yet). I also think the attribute (as opposed to the argument) should declared to be a list, not a sequence, for robustness against future changes which might loop through the steps twice, which could break if we pass in an iterator. To keep things simple, lets just leave the type as List, for now at least?